### ISSUE TITLE
Rationale for storing level and offset in heap nodes?

### ISSUE THREAD

**Issue by dabrahams** [apple/swift-collections#443]
Not really a bug report, but that feature of the implementation surprised me.  A traditional heap stores its elements directly, and as far as I know no heap operations depend on these values.

**Comment by dabrahams** [comment]
You do realize you can compute the level from the index, right?

**Comment by dabrahams** [comment]
> Bit counting is not expensive in absolute terms, but doing that was still measurably slower than this version.

Counting the leading zero bits is one instruction; then you subtract from the bit width. You're telling me memory accesses and the locality costs of extra storage is cheaper?  I must be missing something!

**Comment by lorentey** [comment]
Yes -- counting bits is pretty cheap, but it was still measurably better to have a register hold an incrementally updated level. Saving an integer multiplication's worth of cycles per level touched is still a win: the reference baseline was a regular min-heap that does not need to care about any of this level business.

There are usually no memory accesses or significant locality costs here; `_HeapNode` is materialized on demand, for the duration of an operation. That said, it may be that adding a word to the heap struct that caches the current depth of the tree may be beneficial to `insert`; confirming/rejecting that (and other hypotheses) are left to a future performance pass.

**Comment by dabrahams** [comment]
Oh that's a very interesting design; thanks for the detail!
